src/map/itemdb.cpp

```diff
@@ -2274,194 +2272,396 @@ bool itemdb_isNoEquip(struct item_data *id, uint16 m) {
 	return false;
 }
 
-/**
-* Retrieves random option data
-*/
-struct s_random_opt_data* itemdb_randomopt_exists(short id) {
-	return ((struct s_random_opt_data*)uidb_get(itemdb_randomopt, id));
+const std::string RandomOptionDatabase::getDefaultLocation() {
+	return std::string(db_path) + "/item_randomopt_db.yml";
 }
 
-/** Random option
-* <ID>,<{Script}>
-*/
-static bool itemdb_read_randomopt(const char* basedir, bool silent) {
-	uint32 lines = 0, count = 0;
-	char line[1024];
+/**
+ * Reads and parses an entry from the item_randomopt_db.
+ * @param node: YAML node containing the entry.
+ * @return count of successfully parsed rows
+ */
+uint64 RandomOptionDatabase::parseBodyNode(const YAML::Node &node) {
+	uint16 id;
 
-	char path[256];
-	FILE* fp;
+	if (!this->asUInt16(node, "Id", id))
+		return 0;
 
-	sprintf(path, "%s/%s", basedir, "item_randomopt_db.txt");
+	std::shared_ptr<s_random_opt_data> randopt = this->find(id);
+	bool exists = randopt != nullptr;
 
-	if ((fp = fopen(path, "r")) == NULL) {
-		if (silent == 0) ShowError("itemdb_read_randomopt: File not found \"%s\".\n", path);
-		return false;
+	if (!exists) {
+		if (!this->nodesExist(node, { "Option", "Script" }))
+			return 0;
+
+		randopt = std::make_shared<s_random_opt_data>();
+		randopt->id = id;
 	}
 
-	while (fgets(line, sizeof(line), fp)) {
-		char *str[2], *p;
+	if (this->nodeExists(node, "Option")) {
+		std::string name;
 
-		lines++;
+		if (!this->asString(node, "Option", name))
+			return 0;
 
-		if (line[0] == '/' && line[1] == '/') // Ignore comments
-			continue;
+		if (random_option_db.option_search_name(name)) {
```

Comment:
Should only trigger, if the found id != the current id, otherwise it could be someone redefined the name in the import file.